spring-cloud-commons/src/test/java/reactor/core/publisher/FluxFirstNonEmptyEmittingTests.java

```diff
@@ -0,0 +1,187 @@
+/*
+ * Copyright 2019-2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package reactor.core.publisher;
+
+import java.time.Duration;
+import java.util.Arrays;
+
+import org.junit.Test;
+import org.reactivestreams.Publisher;
+import org.reactivestreams.Subscription;
+import reactor.core.CoreSubscriber;
+import reactor.core.Scannable;
+import reactor.test.StepVerifier;
+
+import static java.util.Collections.singletonList;
+import static org.assertj.core.api.Assertions.assertThat;
+
+/**
+ * @author Tim Ysewyn
+ */
+public class FluxFirstNonEmptyEmittingTests {
+
+	@Test(expected = NullPointerException.class)
+	public void arrayNull() {
+		CloudFlux.firstNonEmpty((Publisher<Integer>[]) null);
+	}
+
+	@Test(expected = NullPointerException.class)
+	public void iterableNull() {
+		CloudFlux.firstNonEmpty((Iterable<Publisher<Integer>>) null);
+	}
+
+	@Test
+	public void firstWinner() {
+		StepVerifier
+				.create(CloudFlux.firstNonEmpty(Flux.range(1, 10), Flux.range(11, 10)))
+				.expectNext(1, 2, 3, 4, 5, 6, 7, 8, 9, 10).verifyComplete();
+	}
+
+	@Test
+	public void firstWinnerSecondEmpty() {
+		StepVerifier.create(CloudFlux.firstNonEmpty(Flux.range(1, 10), Flux.empty()))
+				.expectNext(1, 2, 3, 4, 5, 6, 7, 8, 9, 10).verifyComplete();
+	}
+
+	@Test
+	public void firstWinnerBackpressured() {
+		StepVerifier
+				.create(CloudFlux.firstNonEmpty(Flux.range(1, 10), Flux.range(11, 10)))
+				.thenRequest(5).expectNext(1, 2, 3, 4, 5).thenCancel()
+				.verifyThenAssertThat().hasNotDiscardedElements().hasNotDroppedElements()
+				.hasNotDroppedErrors();
+	}
+
+	@Test
+	public void secondWinner() {
+		StepVerifier
+				.create(CloudFlux.firstNonEmpty(Flux.never(), Flux.range(11, 10).log()))
+				.expectNext(11, 12, 13, 14, 15, 16, 17, 18, 19, 20).verifyComplete();
+	}
+
+	@Test
+	public void secondWinnerFirstEmpty() {
+		StepVerifier
+				.create(CloudFlux.firstNonEmpty(Flux.empty(), Flux.range(11, 10).log()))
+				.expectNext(11, 12, 13, 14, 15, 16, 17, 18, 19, 20).verifyComplete();
+	}
+
+	@Test
+	public void bothEmpty() {
+		StepVerifier.create(CloudFlux.firstNonEmpty(Flux.empty(), Flux.empty()))
+				.expectComplete().verifyThenAssertThat().hasNotDiscardedElements()
+				.hasNotDroppedElements().hasNotDroppedErrors();
+	}
+
+	@Test
+	public void neverAndEmpty() {
```

Comment:
one way to write this test and verify a longer period of time would be to use virtual time. You should also verify that no event occurs, like so:
```java
StepVerifier.withVirtualTime(CloudFlux.firstNonEmpty(Flux.never(), Flux.empty()))
    .expectSubscription()
    .expectNoEvent(Duration.ofDays(1))
    .thenCancel()
    .verifyThenAssertThat()
	.hasNotDiscardedElements().hasNotDroppedElements().hasNotDroppedErrors();
```